classes/ActionScheduler_Abstract_QueueCleaner.php

```diff
@@ -0,0 +1,62 @@
+<?php
+
+/**
+ * Abstract class with common Queue Cleaner functionality.
+ */
+abstract class ActionScheduler_Abstract_QueueCleaner {
+
+	/** @var ActionScheduler_Store */
+	protected $store;
+
+	/**
+	 * Process an individual action.
+	 *
+	 * @param int $action_id The action ID to process.
+	 */
+	protected function process_action( $action_id ) {
+		$action = $this->store->fetch_action( $action_id );
+		try {
+			do_action( 'action_scheduler_before_execute', $action_id );
+			$this->store->log_execution( $action_id );
```

Comment:
Was this intentionally moved outside the `try/catch` block and before the `'action_scheduler_before_execute'` hook? If so, why?

We used to have them run like this back in 2014, but that was changed with SHA: 1da6a46f and SHA: 00c90a3. See #26 and #61 for more info. I think it's better to keep them as they are in v1.5.3 unless there is a compelling reason to go back to this order.

Between these changes and the one where `process_action()` has been made `protected`  (which was changed with SHA: e1a0ab9603d62), I'm wondering if this code was accidentally copied from a much older version of Action Scheduler?

Reply:
It actually was an intentional change to move the `$action = $this->store->fetch_action( $action_id );` statement out of the try/catch block. The main reason for that was that my IDE was giving a warning at the end of the method on the `$this->schedule_next_instance( $action );` line. The warning was that `$action` might not have been defined, since the `try` block could exit earlier (even if that's not likely).

In looking at the flow of `ActionScheduler_wpPostStore::fetch_action()`, I didn't see any possibility of it throwing an exception, so it seemed safe to move that particular statement outside of the try/catch block.

I wasn't aware of the history here, so I can certainly move it back into the block if you think that it would be best to do so.